src/Lean/Server/FileWorker.lean

```diff
@@ -59,24 +59,31 @@ open IO
 open Snapshots
 open JsonRpc
 
+open Widget in
+abbrev StickyDiagnostics := RBTree InteractiveDiagnostic InteractiveDiagnostic.compareAsDiagnostics
+
 structure WorkerContext where
   /-- Synchronized output channel for LSP messages. Notifications for outdated versions are
     discarded on read. -/
-  chanOut          : IO.Channel JsonRpc.Message
+  chanOut              : IO.Channel JsonRpc.Message
   /--
   Latest document version received by the client, used for filtering out notifications from
   previous versions.
   -/
-  maxDocVersionRef : IO.Ref Int
-  hLog             : FS.Stream
-  initParams       : InitializeParams
-  processor        : Parser.InputContext → BaseIO Lean.Language.Lean.InitialSnapshot
-  clientHasWidgets : Bool
+  maxDocVersionRef     : IO.Ref Int
+  /--
+  Diagnostics that are included in every single `textDocument/publishDiagnostics` notification.
+  -/
+  stickyDiagnosticsRef : IO.Ref StickyDiagnostics
```

Comment:
Might it make sense to make this a map `StaleDepId -> InteractiveDiagnostic` instead? It would make for an easier comparison in `handleStaleDependency` (so that you wouldn't need `compareAsDiagnostics`), and you still get to grab all the cached sticky diagnostics quickly by iterating over the values of the map. If other kinds of sticky diagnostics are added later, it could be keyed by `inductive StickyDiagId | staleDep id | ...`.

Reply:
The goal here is specifically to filter sticky diagnostics that look identical (otherwise an `Array` would suffice as well, see also the `fix: deduplicate sticky diagnostics` commit). Without this map, if you save a dependency multiple times, there will be multiple diagnostics in the same location with the same message informing the user that this specific dependency is stale.
Since displaying the same sticky diagnostic multiple times seems like noise in general, I figured it would be best to just implement this at the level of the sticky diagnostics themselves.

Reply:
No I know, but I think the map I proposed would also solve that? If it is keyed by the ID (module name, for instance) of the stale dependency, calling `.insert staleDepName staleDepDiag` would replace the previous entry, not add a new one.

Reply:
I think I'd like to keep it as it is. The sticky diagnostics shouldn't need to know a lot about the diagnostics that are contained in them. I know that we can generalize this further to a sigma type, but I don't really see the point, since I doubt that the sticky diagnostics lookup will ever be a bottleneck.